Public/Src/Engine/Dll/DirectoryScrubber.cs

```diff
@@ -375,5 +372,50 @@ public bool RemoveExtraneousFilesAndDirectories(CancellationToken cancellationTo
                 return true;
             }
         }
+
+        /// <summary>
+        /// Deletes a given array files in parallel.
+        /// Deletion failures are handled gracefully (by logging a warning).
+        /// Returns the number of successfully deleted files.
+        /// </summary>
+        public int DeleteFiles(string[] filePaths)
+        {
+            int numRemoved = 0;
+            using (var timer = new Timer(
+                _ => Tracing.Logger.Log.ScrubbingProgress(m_loggingContext, "", numRemoved, filePaths.Length),
+                null,
+                dueTime: BuildXLEngine.GetTimerUpdatePeriodInMs(m_loggingConfiguration),
+                period: BuildXLEngine.GetTimerUpdatePeriodInMs(m_loggingConfiguration)))
+            {
+                filePaths
+                    .AsParallel()
+                    .WithDegreeOfParallelism(m_maxDegreeParallelism)
+                    .WithCancellation(m_cancellationToken)
+                    .ForAll(path =>
+                    {
+                        if (FileUtilities.FileExistsNoFollow(path) && TryScrubFile(m_loggingContext, path))

```

Comment:
>FileUtilities.FileExistsNoFollow(path) [](start = 28, length = 38)

Do you need to probe beforehand?

Reply:
I don't.  I did it only to get a precise count of deleted files.  If you call our FileUtilities.Delete method for a file that doesn't exists, it will return true.

Reply:
The problem is that `FileUtilities.DeleteFile` is not to be trusted. :) In fact, if you give it a path to a folder, it will delete that folder (maybe only if that folder is empty, but it doesn't matter).

I think probing for file existence here is a small price to pay.